### ISSUE TITLE
[Feature]: Assign a ACL to multiple devices 

### ISSUE THREAD

**Issue by balpoint** [netbox-community/netbox-acls#72]
### NetBox version

v3.3.5

### Feature type

Add a function

### Proposed functionality

Can we give the plugin a option to assign a ACL to multiple devices so that we can assign it to multiple interfaces accross multiple devices if needed. 

### Use case

It will clean up number of acl's that need to be created within the netbox plugin in order to apply a certain ACL to multiple devices.

**Comment by ryanmerolle** [comment]
@abhi1693 please chime in.  I am totally not against this.  I have the same use case.

BUT its not truly how the actual model works.  @balpoint may be really strict on ACL names and ACL definitions, but that is not always the case.  You could say "well then people should not associate that ACL to another device."  But what do we do if someone intends to use the same exact ACL defintion on 2 different devices, then their peer or their replacement comes along and wonders why they cannot edit/split the ACL so that the ACL is now 2 different ACLs (one per device).

Long story short, if we do this, we may want a feature to allow users to "fork" the ACL thus breaking the relationship, but cloning all the ACL rule lines.

I guess this is a bigger can of worms, we need to think through is my point.

**Comment by abhi1693** [comment]
One way could be to do it like `Contacts` assignment to objects work. Basically, create another model which works like a glue to between the `AccessList` and related objects. Although, I'm not sure if that's the path we should take at the moment.

**Comment by ryanmerolle** [comment]
That's not a bad idea, but we would need to think about it some more.  It's honestly more of an abstraction that does not exist on its own in a network.

**Comment by Kani999** [comment]
I've run into the same problem as @balpoint 
We have an access list used on almost every device with the same name. 
It's a little messy when a lot of "duplicates" is stored in the plugin. 

But I guess I could use something as `context_data` for this.

**Comment by ryanmerolle** [comment]
I’m game for adding to the models for an abstraction.  

We just need some ideas to allow users to define something like a  standardized multi-device ACL.  We also need a way to clone one of these ACLs to a single device ACL so that users can model the real world when they need to edit an ACL on a single device to be outside of the “standard.”

This sort of abstraction is not found anywhere currently in netbox or many (if any plugins) besides like config context.  I’ll have some free time in August to think about this some more.  In the meantime I’d be game for ideas, a draft of the model update, and even a PR.

**Comment by ryanmerolle** [comment]
The other approach could be to implement a template like approach.  For example similar to devices and devicetypes.  The template would apply only at creation.

What are everyone’s thoughts for the two approaches?

**Comment by Kani999** [comment]
The approach of "template-like" solutions does not seem right to me. It would create a massive amount of records on my side. 
For example, if I have the same access list on 100 devices with 20 rules in them it would create 2000 standard rules which would be always the same. The editing would be a little bit of a pain (of course not via API).
There is a problem with the UI. If all "common access lists" are named the same, when it comes to selecting the ACL when assigning it to the interface it looks like this: 

![image](https://github.com/ryanmerolle/netbox-acls/assets/17337078/08993aca-bbfd-47ab-b6e4-0f714508d5d5)


I like the idea of an @abhi1693 that suggests using an "access list assignment" model in which you could assign an ACL to a Device, Interface, or other models. 

I think when it comes to "splitting" the common ACL like you described:
> But what do we do if someone intends to use the same exact ACL definition on 2 different devices, then their peer or their replacement comes along and wonders why they cannot edit/split the ACL so that the ACL is now 2 different ACLs 

But when you use "common access list" then the logic of "oh I will just edit this rule on this particular device" disrupts the logic of the "common access-list" and shouldn't be done.  But If it comes, the common access list should be removed from the device and a new one (specific for the device) will be created which will have different rules in it. 
 If there would be an option to copy/clone the existing ACLs with the rules, it would be a great help.
 
 That's just my opinion, maybe I'm missing something.

**Comment by ryanmerolle** [comment]
### Option 1
@abhi1693 and I talk on slack a bit about this.  The method he described is the first options I was poorly describing.  Its a an approach that would **STILL** allow user to granularly map 1 ACL to device to model the real world because 1 device as no understanding of another device's ACL definitions, **BUT ALSO** introduce a new model to share across devices as your abstraction layer.  From a user perspective this makes sense for me for the same reasons you mentioned @Kani999.  It is still confusing because you would have 2 different types of associations.

#### Option 1 Variation
Another variation of that approach could be to edit the existing model and just have some sort of boolean/checkbox to signify the type of ACL (device or global).

### Option 2
The template model is not ideal if you update NetBox using the UI only, but there is a few reasons why people might prefer the template model.
1. ACLs can be generated from the current template just like the real world.  I have a design, and I apply at first build.
2. Like the real world, you can model the current state of ACLs applied to device
3. You can build scripts (we can even add to the plugin) that would show you where your applied ACLs differ from a template.

TLDR for this option is that it models the real world just like most of the NetBox core models.  IE NetBox core is currently not a SDN controller that abstracts this sort of logic (for better or worse).  The exception to this is really models like VLANs and config_context.



### Final Thoughts

- We can hammer out the above and set a poll up.  
- I am totally not biased to any option, just trying to represent them all (ie I would represent option 1 more if you said you liked option 2).
- I think NetBox/Nautobot/Device42/etc all do not handle abstractions well in their current state.

**Comment by Kani999** [comment]
@ryanmerolle Great summarization.
I would prefer option 1 from my point of view.

**Comment by balpoint** [comment]
Hello everyone,

As the person who initiated the feature request, I wanted to join the discussion again. It's fantastic to see the community coming together and brainstorming ways to improve this plugin with a wider range of methods. Kudos to everyone involved!

Regarding the ongoing discussion, I want to thank @ryanmerolle for summarizing the options discussed outside of the feature request. In my opinion, option 1 seems the most convenient and straightforward. By simply checking a box, we can make an Access Control List (ACL) global, allowing it to be applied to multiple interfaces across multiple devices. This approach would provide a one-to-many status for the ACL, streamlining its usage.

**Comment by cs-1** [comment]
Thanks for looking into this feature request. We have another use case for this: in fabric constructs (VXLAN + EVPN-BGP) you have L3 router interfaces for the same LAN on different machines but need to make sure that all ACLs are in sync. It would be great to be able to assign the same ACL to different machines. So I definitely second this FR.

**Comment by aoforb** [comment]
We have several thousand routers and switches with four standard ACLs. 

- VTY line ACL allowing only hand full of management sources
- RP multicast address
- SNMP ACL
- NTP ACL

When I installed this plugin today, I was expecting to maintain only four ACLs on Netbox. The current model would mean MANY duplicated ACLs would need updating when the standard changes.

Glad to see some discussion on the topic, how far away is the solution?

**Comment by ghost** [comment]
This was more or less the initial feedback from my networking guys, that documenting the current state is one thing, but planning the intended state is also important.

From that point of view, you'd like to have ACLs that are not associated with any device, which you can build and then apply to multiple devices. An alternative would be to manage device groups, and give the ACLs e.g. a "planned" state (such as some other core functionality of netbox offers).

**Comment by ianbarrere** [comment]
I am also in favor of this change. All our edge routers use the same access-list layout, with only differing object-groups (support for the latter requested in [issue 147](https://github.com/netbox-community/netbox-acls/issues/147) and [56](https://github.com/netbox-community/netbox-acls/issues/56)), so it would be nice to be able to account for that in the plugin.

**Comment by pheus** [comment]
**Proposal: Assign a single ACL to multiple objects (devices, interfaces, VMs)**

_First off, thanks for the thoughtful discussion and for keeping the plugin grounded in how ACLs work on real devices. I’d like to offer an additional approach that keeps the “one device = independent ACL definition” model intact, while making it easier to reuse the same ACL across many targets when that’s desirable._

### Goals

- Keep today’s 1:1 realism available (an ACL can be scoped to a single device/VM just as it is now).
- Add a clean way to reuse the same ACL across multiple devices/VMs and/or interfaces.
- Minimize user confusion by unifying assignment concepts and keeping the UI predictable.

---

## Option A — Add **DeviceGroup** as a reusable target

Introduce a reusable grouping primitive and allow an `AccessList` to be assigned to the group.

**What users get**
- Create a `DeviceGroup` once (devices and/or VMs), then attach any number of ACLs to it.
- Continue to assign ACLs directly to a single Device/VM when they need per‑device uniqueness/variance.

**Sketch**

```python
# New, reusable grouping primitive (not tied to a single ACL)
class DeviceGroup(NetBoxModel):
    name = models.CharField(max_length=100, unique=True)
    # optional: slug, description, tenant/tenant_group
    devices = models.ManyToManyField("dcim.Device", blank=True)
    virtual_machines = models.ManyToManyField("virtualization.VirtualMachine", blank=True)
```

Then allow `AccessList` assignments to include `DeviceGroup` (see Option C below for a unified assignment model).

**Pros**
- Reuse without duplicating ACLs.
- Groups are reusable across multiple ACLs.
- Mirrors how many teams plan rollouts (“apply to all edge routers” etc).

**Trade‑offs**
- Adds a new concept to the UI.

---

## Option B — Extract **ACL–Object** assignment into its own model (no new grouping)

Decouple the per‑object assignment from the `AccessList` itself. This enables the same ACL to be assigned multiple times (different targets) without changing how ACLs are authored.

**Sketch**

```python
class ACLObjectAssignment(NetBoxModel):
    access_list = models.ForeignKey(
        "netbox_acls.AccessList",
        on_delete=models.CASCADE,
        related_name="object_assignments",
    )

    assigned_object_type = models.ForeignKey(
        to=ContentType,
        on_delete=models.PROTECT,
        limit_choices_to=ACL_OBJECT_ASSIGNMENT_MODELS,
        verbose_name=_("Assigned Object Type"),
    )
    assigned_object_id = models.PositiveBigIntegerField()
    assigned_object = GenericForeignKey("assigned_object_type", "assigned_object_id")

    # Optional: denormalized for name-collision checks (see "Uniqueness" below)
    _acl_name = models.CharField(max_length=100, editable=False, db_index=True)


ACL_OBJECT_ASSIGNMENT_MODELS = (
    Q(app_label="dcim", model__in=("device")) |
    Q(app_label="virtualization", model__in=("virtualmachine"))
)
```

**Pros**
- No new grouping concept; just “assign the same ACL to more things.”
- Keeps the “realism” path intact by still allowing one‑off device assignments.

**Trade‑offs**
- Users still need to pick many devices/VMs individually (bulk actions help).

---

## Option C — Unify interface and object assignments into one model

Refactor `ACLInterfaceAssignment` into the generalized `ACLObjectAssignment` so there’s one place and one mental model for “where is this ACL applied?”

```python
class ACLObjectAssignment(NetBoxModel):
    access_list = models.ForeignKey(
        "netbox_acls.AccessList",
        on_delete=models.CASCADE,
        related_name="object_assignments",
    )

    assigned_object_type = models.ForeignKey(
        to=ContentType,
        on_delete=models.PROTECT,
        limit_choices_to=ACL_OBJECT_ASSIGNMENT_MODELS,
        verbose_name=_("Assigned Object Type"),
    )
    assigned_object_id = models.PositiveBigIntegerField()
    assigned_object = GenericForeignKey("assigned_object_type", "assigned_object_id")

    # Required for interfaces only; optional/empty for device/VM
    direction = models.CharField(
        max_length=12,
        choices=(("ingress","Ingress"),("egress","Egress")),
        blank=True,
        null=True,
    )

    # Optional: denormalized for name-collision checks (see "Uniqueness" below)
    _acl_name = models.CharField(max_length=100, editable=False, db_index=True)


ACL_OBJECT_ASSIGNMENT_MODELS = (
    Q(app_label="dcim", model__in=("device","interface")) |
    Q(app_label="virtualization", model__in=("virtualmachine","vminterface"))
)
```

**Behavior**
- If `assigned_object` is an `Interface`/`VMInterface`: `direction` is required.
- If `assigned_object` is a `Device`/`VirtualMachine`: `direction` is not used.

**Why this helps**
- One assignment table, one set of API endpoints/filters, one UI for bulk actions.
- Fewer concepts to explain in docs.

---

## Uniqueness, validation, and constraints

**What we must prevent**
1. The same ACL assigned twice to the exact same target (device/VM/interface + optional direction).
2. (Optional policy) The same _ACL name_ appearing more than once on the same target via different `AccessList` rows.

**Suggested approach**
- DB‑level unique constraint for `(access_list, assigned_object_type, assigned_object_id, direction)`.
  - For device/VM rows where `direction` is not used, enforce uniqueness at the application layer (`clean()`), or use a sentinel value (e.g., `"none"`) rather than `NULL` so the DB constraint applies uniformly.
- If you want policy (2), keep `_acl_name` denormalized and add a **second** unique constraint on `(_acl_name, assigned_object_type, assigned_object_id, direction/sentinel)`. Add a signal or `save()` override to keep `_acl_name` in sync when `access_list.name` changes.
  - This is optional but matches the “on a device an ACL name should be unique” expectation many teams have.

---

## API & UI considerations

- **AccessList creation**: lift the requirement to pick a target at creation; authors can create the ACL first, then add one or more assignments.
- **AccessList detail page**: show a tabular “Assignments” section with object type, object, direction, and created/updated.
- **Bulk actions**: “Assign to… (devices/VMs/interfaces/groups)” and “Unassign” bulk actions will be key to usability.
- **Filters**: allow filtering assignments by object type, object ID, site, location, tenant, platform, tag, and direction.
- **API**: expose `/access-lists/{id}/assignments/` as a nested endpoint; support bulk create/delete of assignments.

---

## Migration path (non‑breaking)

1. **Data migration**: For each existing `AccessList` that currently points directly at a Device/VM (or interface via `ACLInterfaceAssignment`), create corresponding `ACLObjectAssignment` rows.
2. **Tests**: Add unit tests covering duplicate prevention, interface direction requirements, and migration parity.

---

## Open questions

- Should “ACL name must be unique per device/VM” be enforced (via `_acl_name`) or left to user policy?

---

### TL;DR
- **Minimum change** to gain reuse: Option **B**, add `ACLObjectAssignment`.
- **Best UX** over time: Option **C**, unify interface & object assignments so there’s a single, consistent assignment model.
- **Optional enhancement**: Option **A**, `DeviceGroup` for easy many‑target assignments and reuse.

Thanks again for maintaining this plugin and for the thoughtful discussion on scope vs. abstraction. If there’s alignment on the direction, I’m happy to help with a draft model + migration PR.

**Comment by ghost** [comment]
@pheus thank you, this is an excellent summary!

Just going by your description, I'd favour option C.

**Comment by alehaa** [comment]
I would also choose option C. Lifting the requirement to assign the list would also help with #268. Just some thoughts:

* I would make the ACL unique per device. Of course, one shouldn't duplicate assignments, but having two different ACLs with the same name wouldn't work, at least for Cisco IOS and other similar operating systems, even if they are assigned to different interfaces/directions. Implementing a check in the `clean()` method might be necessary to solve this issue.
* When assigning an ACL to a device, we could mark it as inbound. I was thinking this mapping could be used for SSH access, for example, and it would be inbound, right?
* I’m not familiar with nested API endpoints. It might be easier to use a separate endpoint and filters, just like contact assignments do.

**Comment by pheus** [comment]
Thank you very much for your thoughts on this topic.

@alehaa, I agree that we should still make the ACL assignment unique per device, although this means the uniqueness of the ACL `name` can’t be enforced on the `AccessList` model. Adding the cached ACL name `_acl_name` from the assigned ACL can enforce uniqueness at the DB level, whereas the `clean()` method only catches it at the application level (which might be fine, after all).

I understand your point regarding the inbound direction for the device assignment. But, as you probably know, there are other uses of ACLs that are not bound to a VTY or interface. In addition, a third choice in the DB can be handled differently in the UI (e.g., N/A). I’d suggest a tri‑state for `direction`: `ingress | egress | none`. I’ll test this during implementation. :)

I agree with the last point. Although I’m a fan of the nested URL schema, as it provides context, it would lead to an unnecessary breaking change to the interface-assignment API.

Thanks again, everyone.

**Comment by pheus** [comment]
Thanks everyone for the thoughtful input - just a quick update and a few questions.

### Update
I’m currently working on *Variant C*, which uses a single model, `ACLAssignment`, for both **host** (Device, VirtualChassis, VirtualMachine) and **interface** (Interface, VMInterface) assignments. The main assignment flow works as expected, including the related views (`ChildrenViews`) on all supported models.

For **host assignments**, the `direction` field is disabled in forms and stored as `"none"`.

During implementation I decided to follow @alehaa’s suggestion to enforce **name uniqueness** via validation logic only (instead of caching a denormalized name). The cached-field route adds considerable complexity when handling ACL renames.

---

### Open questions (feedback appreciated)

1. **Host prerequisite for interface assignments**  
   Should assigning an ACL to a **host** be required before it can be assigned to an **interface**?  
   - It’s currently implemented as a requirement (in v1.9.0).  
   - If “yes,” I’m inclined to split host and interface assignments into **separate models** to keep the rules clear.

2. **Renaming ACLs with existing assignments**  
   Should ACLs be renameable when assignments exist?  
   - If allowed, is it sufficient to validate conflicts only against **host assignments** (works in current WIP), or should we also validate **interface assignments**? (Depends on the answer to 1.)

3. **ACL name uniqueness & Web UI clarity**  
   If we **allow multiple ACLs with the same name**, the NetBox **Web UI cannot reliably distinguish them** (lists, lookups, bulk actions).  
   - One option is to keep the ACL `name` globally unique in NetBox.  
   - Alternatively, use two fields: a unique NetBox reference `name` and a `host_acl_name` used on the target host (e.g., `name = "Management ACL (all devices)"`, `host_acl_name = "mgmt"`), and display both to disambiguate.

Thanks again for your time and guidance!

**Comment by alehaa** [comment]
I’ll keep this short. Here’s my opinion on the questions:

1. Interface assignments should be sufficient and save an additional step, making management simpler in daily workflows.
2. Renaming ACLs should be allowed. While I understand that this alters device configuration, it’s the same "problem" for VLANs, interfaces, and so on. Since NetBox core doesn’t implement immutable objects, neither should we.
3. Making ACL names globally unique would significantly simplify the implementation. If there’s a need for multiple ACLs with the same name, this could be added later.

**Comment by pheus** [comment]
Thanks, much appreciated!

I agree with your points; however, point 3 is tricky. I have to assume some users have multiple ACLs with the same name assigned to different devices. I can’t safely migrate these datasets automatically, so I think we need to allow duplicate names for now.

To help disambiguate in the UI, I can include the DB ID in the `__str__` representation (e.g., `name [id: 123]`). Do you have other ideas or suggestions?

Thanks!

**Comment by aoforb** [comment]
Do you have an idea when this would be merged? 

Looks great!

Thanks

**Comment by pheus** [comment]
Hi @aoforb.  Thanks for checking in!

I can’t speak for the maintainers, but this is part of the larger, potentially breaking 2.0 release. Since we’re quite late in the NetBox release cycle, I assume this will target v4.5, which is scheduled for January 2026.
